lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,94 @@
+class LRUHash
+  # This LRU (Least Recently Used) hash will allow
+  # the cleaning of resources as time goes.
+  # The  goal is to remove the least recently used resources
+  # everytime we set a new resource. A default window of
+  # 5 minutes will allow empty item to stay in the hash
+  # for a maximum of 5 minutes
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?
+  attr_reader :table, :minimum_time_in_lru
+  MINIMUM_TIME_IN_LRU = 300
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+
+    def try_lock
+      true
+    end
+
+    def unlock
+      true
+    end
+  end
+
+  [:keys, :clear].each do |attribute|
+    define_method :"#{attribute}" do
+      @lock.synchronize { @table.public_send(attribute) }
+    end
+  end
+
+  def initialize
+    @minimum_time_in_lru = MINIMUM_TIME_IN_LRU
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, resource)
+    @lock.synchronize do
+      @table.delete(key)
+      @table[key] = resource
+    end
+    resource.updated_at = Time.now
```

Comment:
I can't think of why setting `resource.updated_at` outside the mutex is unsafe. That being said, this feels weird because we're writing to a shared resource without a lock. Feel free to mark this comment as resolved without changing anything.